src/test/test-audit-util.c

```diff
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "tests.h"
+#include "audit-util.h"
+
+TEST(audit_loginuid_from_pid) {
+        _cleanup_(pidref_done) PidRef self = PIDREF_NULL, pid1 = PIDREF_NULL;
+        int r;
+
+        assert_se(pidref_set_self(&self) >= 0);
+        assert_se(pidref_set_pid(&pid1, 1) >= 0);
+
+        uid_t uid;
+        r = audit_loginuid_from_pid(&self, &uid);
+        assert_se(r >= 0 || r == -ENODATA);
```

Comment:
This is not a runtime environment, it's a unit test, it can run in a chroot with nothing in it apart from compiler and libraries

Reply:
> This is not a runtime environment, it's a unit test, it can run in a chroot with nothing in it apart from compiler and libraries

yeah, we don't support that though. all our tests will fail.

is this a real thing? is debian build system not mounting /proc/? whats going on?

> if (proc_mounted() == 0)
>    return -ENOSYS;

this is paranoia for cases where we ourselves set up some incomplete sandbox or are in the process if setting it up and haven't gotten everything right yet. It's *not* support for allowing CIs to run without /proc/.

Sorry, but half of Linux' process API is in /proc/. You cannot reasonably run anything non-trivial without it, and we *never* supported that, and made this very very clear.

Reply:
I mean, the CI passes, so I think we are all good.

Reply:
> yeah, we don't support that though. all our tests will fail.

No this is not true. Integration tests, runtime stuff - sure. But unit tests? Nope. Once again, unit tests are _supposed_ to run during builds, which happen in the most locked down, isolated, weird and diverse environments. It's fine that a unit test tries to do stuff, but it _must_ fail gracefully when there are permission, access or other environmental issues. It's absolutely ok to `return log_tests_skipped()` when things can't work, that's why it exists in the first place.

For things that _must_ run in prescribed environments, there are integration tests, where we set up anything we want, in the way we want it. And for special stuff that is meant only for your laptop, there's the "manual" tests list defined in meson, it's fine to add stuff there too, that gets run only manually. But `meson test` needs to run anywhere.

Reply:
> I mean, the CI passes, so I think we are all good.

that's because we don't have good coverage of such environments, at some point I'll add such a job

Reply:
Uh, sorry, but environments without /proc are bogus. Sorry. I have never seen anything like this, and I don#t think we need to care. We never cared, and it's really the wrong time to begin with that. I am absolutely happy if all our tests fail in such an idiotic environment...

I mean, where is this coming from? where is there a CI system that doesn't mount /proc?

Reply:
> Uh, sorry, but environments without /proc are bogus.

Once again, this is confusing runtime environments with build environments. The latter _only_ need a compiler toolchain, libraries and headers, and that's it - `meson test` needs to work in those conditions. Runtime environments are of course different. Think about it for a second: it's normal that the build environment and the target runtime environment differ wildly, as we have seen for the BTF header case. It's perfectly normal to run a build process on a 10 years old kernel or so, that builds packages that will be installed and ran on the bleeding edge. What sense does it make to assert on the kernel API behaviour in such a situation? It doesn't represent the target environment, and it's not supposed to, it's a build environment.

Reply:
@bluca I really wouldn't die on this hill, we have hundreds of other violations of needing /proc in the unit tests that will fail without it available. And extremely likely to regress unless we get coverage upstream for the environment.

Missing /sys is fine of course and already covered by the mkosi builds since /sys isn't available there.

Reply:
I am not confusing anything. A build environment needs /proc/. And before we continue the discussion, name me *one* real-life Linux build environment that doesn't mount /proc/.

You are making things up, there is no such thing. Not even debian does that.

Reply:
A build environment does not need proc, nor sys, nor dev, it just needs a compiler toolchain and libraries and such build deps, because it's a build environment, not a runtime environment. Again, it's perfectly normal to have 10 years old LTS kernels in build envs, so such expectations of what interfaces the kernel should provide cannot be met _when building packages_, they can only be met in the target runtime environment. It's trivial to create such an environment, you just need deboostrap and chroot, you don't even need wrapper scripts, although they do exist, such as pbuilder and family iirc.

I mean, even systemctl and other _runtime_ tools work with reduced functionality when there's no proc, so that you can do 'systemctl enable' in such environments when images are built/finalized. With a big fat warning of course (which is fine!), but they do work:

```
# ls /proc/
# build/systemctl enable foo
⚠️ /proc/ is not mounted. This is not a supported mode of operation. Please fix
your invocation environment to mount /proc/ and /sys/ properly. Proceeding anyway.
Your mileage may vary.
Created symlink '/etc/systemd/system/multi-user.target.wants/foo.service' → '/etc/systemd/system/foo.service'.
```

I've just checked, and out of 1700 unit tests, 156 fail in such an env. Out of those 156, about half (77) are due to a single shell script assuming it can use /dev/, and it's a one line fix to skip. I'll look into the other 79, provide fixes and a CI job for this.

Reply:
Ok, you cannot name any.

Sorry but those non-existing setups really shouldn't run our test suite. It would not yield useful information, anyway. You are welcome to waste your time on such hypothetics, but please dont waste mine.

Reply:
This should be addressed in #35101. Of course, it needs to be addressed in this PR if #35101 is merged earlier. But, currently, we do not have any consensus about #35101. It is too early to discuss that in this PR.

Reply:
Yes this can go in first